File: Binary Search Tree/352. Data Stream as Disjoint Intervals.cpp
/*
352. Data Stream as Disjoint Intervals

Given a data stream input of non-negative integers a1, a2, ..., an, ..., 
summarize the numbers seen so far as a list of disjoint intervals.

For example, suppose the integers from the data stream are 1, 3, 7, 2, 6, ..., then the summary will be:

[1, 1]
[1, 1], [3, 3]
[1, 1], [3, 3], [7, 7]
[1, 3], [7, 7]
[1, 3], [6, 7]
Follow up:
What if there are lots of merges and the number of disjoint intervals are small compared to the data stream's size?

*/
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class SummaryRanges {
public:
    /** Initialize your data structure here. */
    SummaryRanges() {}
    vector<Interval> vec;
    
    void addNum(int val) {
        auto cmp = [](Interval a, Interval b){return a.start<b.start;};
        auto it = lower_bound(vec.begin(),vec.end(),Interval(val, val), cmp);
        if (it!=vec.begin() && (it-1)->end >= val-1) it--;
        int start = val, end = val;
        while(it!=vec.end() && val+1>=it->start && val-1<=it->end){
            """
            比如 1,3,7,2 结果是[1,3],[7,7], 但是[1,4,7,2] 结果是[1,2],[4,4],[7,7]
            2+1 >= 1 && 2-1 < = 1
            2 + 1>=3 && 2-1 <= 3
            """

            start = min(start, it->start);
            end = max(end, it->end);
            it = vec.erase(it);
        }
            
        vec.insert(it,Interval(start, end));
    }
    
    vector<Interval> getIntervals() {
        return vec;
    }
};

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges obj = new SummaryRanges();
 * obj.addNum(val);
 * vector<Interval> param_2 = obj.getIntervals();
 */


class SummaryRanges {
public:
    /** Initialize your data structure here. */
    SummaryRanges() {}
    vector<Interval> vec;
    
    void addNum(int val) {
        auto cmp = [](Interval a, Interval b){return a.start<b.start;};
        auto it = lower_bound(vec.begin(),vec.end(),Interval(val, val), cmp);
        if (it!=vec.begin() && (it-1)->end >= val-1) it--;
        int start = val, end = val;
        while(it!=vec.end() && end+1>=it->start){
            start = min(start, it->start);
            end = max(end, it->end);
            it = vec.erase(it);
        }
            
        vec.insert(it,Interval(start, end));
    }
    
    vector<Interval> getIntervals() {
        return vec;
    }
};


/*

using Set
*/



class SummaryRanges {
public:
    /** Initialize your data structure here. */
    SummaryRanges() {}
    
    void addNum(int val) {
        auto it = st.lower_bound(Interval(val, val));
        if (it!=st.begin() && (--it)->end < val-1) it++;
        int start = val, end = val;
        while(it!=st.end() && val+1>=it->start && val-1<=it->end){
            start = min(start, it->start);
            end = max(end, it->end);
            it = st.erase(it);
        }
            
        st.insert(it,Interval(start, end));
    }
    
    vector<Interval> getIntervals() {
        vector<Interval> vec;
        for(auto i: st){
            vec.push_back({i.start,i.end});
        }
        return vec;
    }
    
    struct Cmp{
        bool operator()(Interval a, Interval b){ return a.start < b.start; }
    };
    set<Interval, Cmp> st;
};




// Time:  addNum: O(logs), getIntervals: O(s), s is the data stream's size.
// Space: O(s)
class SummaryRanges {
public:
    /** Initialize your data structure here. */
    SummaryRanges() {
        
    }
    
    void addNum(int val) {
        nums_.emplace(val);
    }
    
    vector<Interval> getIntervals() {
        vector<Interval> result;
        if (nums_.empty()) {
            return result;
        }
        auto start = *nums_.begin(), end = *nums_.begin();
        for (auto it = next(nums_.begin()); it != nums_.end(); ++it) {
            if ( *it == end + 1) {
                end = *it;
            } else  {
                result.emplace_back(start, end);
                if (it != nums_.end()) {
                    start = end = *it;
                }
            }
        }
        result.emplace_back(start, end);
        return result;
    }

private:
    set<int> nums_;
};


/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class SummaryRanges {
public:
    void addNum(int val) {
        auto Cmp = [](Interval a, Interval b) { return a.start < b.start; };
        auto it = lower_bound(vec.begin(), vec.end(), Interval(val, val), Cmp);
        int start = val, end = val;
        if (it!=vec.end()) 
            cout<<(it->start)<< "  end  "<<(it->end)<<endl;
        if(it != vec.begin() && (it-1)->end+1 >= val) {
            cout<<"  it -- "<<" (it-1)->end "<<(it-1)->end<<"  val "<<val<<endl;
            it--;
        }
        while(it != vec.end() && val+1 >= it->start && val-1 <= it->end) 
        //比如 1,3,7,2 结果是[1,3],[7,7], 但是[1,4,7,2] 结果是[1,2],[4,4],[7,7]
        {
            cout<<" val "<<val<<" start "<<start<< " it->start "<<it->start<<" end "<<end<<" it->>edn "<<it->end<<endl;
            start = min(start, it->start);
            end = max(end, it->end);
            it = vec.erase(it);
            cout<<" after start "<<start<<" end "<<end<<endl;
        }
        cout<<" insert "<<val<<" start "<<start<<" end "<<end<<endl;
        vec.insert(it,Interval(start, end));
    }
    
    vector<Interval> getIntervals() {
        return vec;
    }
private:
    vector<Interval> vec;
};
/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges obj = new SummaryRanges();
 * obj.addNum(val);
 * vector<Interval> param_2 = obj.getIntervals();
 */

File: Binary Search Tree/352. Data Stream as Disjoint Intervals.py
"""
352. Data Stream as Disjoint Intervals

[1, 1]
[1, 1], [3, 3]
[1, 1], [3, 3], [7, 7]
[1, 3], [7, 7]
[1, 3], [6, 7]
"""
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class SummaryRanges(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.heap = []

    def addNum(self, val):
        """
        :type val: int
        :rtype: void
        """
        if val not in (x[0] for x in self.heap):
            heapq.heappush(self.heap, (val, Interval(val, val)))

    def getIntervals(self):
        """
        :rtype: List[Interval]
        """
        stack = []
        while self.heap:
            val, cur = heapq.heappop(self.heap)
            if not stack:
                stack.append((val,cur))
            else:
                _, prev = stack[-1]
                if cur.start <= prev.end+1:
                    prev.end = max(prev.end, cur.end)
                else:
                    stack.append((val,cur))
        self.heap = stack
        return list(map(lambda x: x[1], stack))

        


# Your SummaryRanges object will be instantiated and called as such:
# obj = SummaryRanges()
# obj.addNum(val)
# param_2 = obj.getIntervals()

class SummaryRanges:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.dic = []
        

    def addNum(self, val):
        """
        :type val: int
        :rtype: void
        """
        def upper_bound(target):
            left, right = 0, len(self.dic)-1
            while left<=right:
                mid = int(left + (right-left)/2)
                if self.dic[mid].start>target:
                    right = mid - 1
                else:
                    left = mid + 1
            return left 
        i = upper_bound(val)
        if i!=0 and self.dic[i-1].end+1>=val:
            i-=1
        start, end = val, val
        while i<len(self.dic) and end+1>=self.dic[i].start: #现在的结束加1，大于下一个开始，比如[1,2] 和[3,3] 要结合成[1,3]
            end = max(end, self.dic[i].end)
            start = min(start,self.dic[i].start)
            del self.dic[i]
        self.dic.insert(i,Interval(start, end))

    def getIntervals(self):
        """
        :rtype: List[Interval]
        """
        return self.dic



class SummaryRanges:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.dic = []
        

    def addNum(self, val):
        """
        :type val: int
        :rtype: void
        """
        def upper_bound(target):
            left, right = 0, len(self.dic)-1
            while left<=right:
                mid = int(left + (right-left)/2)
                if self.dic[mid].end>target:  #用end 比较也可以，想找到点离target最近的
                    right = mid - 1
                else:
                    left = mid + 1
            return left 
        i = upper_bound(val)
        if i!=0 and self.dic[i-1].end+1>=val:
            i-=1
        start, end = val, val
        while i<len(self.dic) and end+1>=self.dic[i].start:
            end = max(end, self.dic[i].end)
            start = min(start,self.dic[i].start)
            del self.dic[i]
        self.dic.insert(i,Interval(start, end))

    def getIntervals(self):
        """
        :rtype: List[Interval]
        """
        return self.dic
